src/reanimated2/animation/transformationMatrix/matrixUtils.tsx

```diff
@@ -0,0 +1,369 @@
+type FixedLengthArray<
+  T,
+  L extends number,
+  PassedObject = [T, ...Array<T>]
+> = PassedObject & {
+  readonly length: L;
+  [I: number]: T;
+};
+
+export type AffiniteMatrix = FixedLengthArray<FixedLengthArray<number, 4>, 4>;
+
+export type AffiniteMatrixFlat = FixedLengthArray<number, 16>;
+
+export type TransformMatrixDecomposition = Record<
+  | 'translationMatrix'
+  | 'scaleMatrix'
+  | 'rotationMatrix'
+  | 'skewMatrix'
+  | 'scaleMatrix',
+  AffiniteMatrix
+>;
+
+interface TansformMatrixDecompositionWithAngles
+  extends TransformMatrixDecomposition {
+  rx: number;
+  ry: number;
+  rz: number;
+}
+
+export function isAffiniteMatrixFlat(x: unknown): x is AffiniteMatrixFlat {
+  'worklet';
+  return (
+    Array.isArray(x) &&
+    x.length === 16 &&
+    x.every((element) => typeof element === 'number')
+  );
+}
+
+export function isAffiniteMatrix(x: unknown): x is AffiniteMatrix {
+  'worklet';
+  return (
+    Array.isArray(x) &&
+    x.length === 4 &&
+    x.every(
+      (row) =>
+        Array.isArray(row) &&
+        row.length === 4 &&
+        row.every((element) => typeof element === 'number')
+    )
+  );
+}
+
+export function flatten(matrix: AffiniteMatrix): AffiniteMatrixFlat {
+  'worklet';
+  return matrix.flat() as AffiniteMatrixFlat;
+}
+
+export function unflatten(matrix: AffiniteMatrixFlat): AffiniteMatrix {
+  'worklet';
+  const result = [];
+  const matrixCopy = [...matrix];
+  while (matrixCopy.length) {
+    result.push(matrixCopy.splice(0, 4));
+  }
+  return result as AffiniteMatrix;
+}
+
+function maybeFlattenMatrix(
+  matrix: AffiniteMatrix | AffiniteMatrixFlat
+): AffiniteMatrixFlat {
+  'worklet';
+
+  return isAffiniteMatrix(matrix) ? flatten(matrix) : matrix;
+}
+
+export function multiplyMatrices(
+  a: AffiniteMatrix,
+  b: AffiniteMatrix
+): AffiniteMatrix {
+  'worklet';
+  const output: AffiniteMatrix = [
+    [0, 0, 0, 0],
+    [0, 0, 0, 0],
+    [0, 0, 0, 0],
+    [0, 0, 0, 0],
+  ];
+
+  for (let x = 0; x < 4; x++) {
+    for (let y = 0; y < 4; y++) {
+      for (let i = 0; i < 4; i++) {
+        output[x][y] += a[x][i] * b[i][y];
+      }
+    }
+  }
+
+  return output;
+}
+
+function linearOperationOnMatrix<T extends AffiniteMatrixFlat | AffiniteMatrix>(
+  _a: T,
+  func: (x: number) => number
+): T {
+  'worklet';
+
+  const isFlatOnStart = !!isAffiniteMatrixFlat(_a);
+  const a: AffiniteMatrixFlat = maybeFlattenMatrix(_a);
+
+  a.forEach((_, i) => {
+    a[i] = func(a[i]);
+  });
+
+  return isFlatOnStart ? (a as T) : (unflatten(a) as T);
+}
+
+function linearOperationOnMatrices<
+  T extends AffiniteMatrixFlat | AffiniteMatrix
+>(_a: T, _b: T, operation: 'add' | 'subtract'): T {
```

Comment:
According to the convention, underscore (`_`) at the beginning of the argument name means that the argument is unused, however, it's not the case here.